### PR TITLE
[Calyx] [SCFToCalyx] Add collapse-control canonicalization pattern for IfOp, SeqOp and ParOp.

### PR DESCRIPTION
**Pull request by cgyurgyik**
Adds some canonicalization methods for SeqOp and ParOp to replace the pattern rewrite in SCFToCalyx.. This is similar to the native compiler's [Collapse Control](https://github.com/cucapra/calyx/blob/master/calyx/src/passes/collapse_control.rs) pass. 

For the cells section, uses `CHECK-DAG` since we don't particularly care what order the cells are listed in. This was previously causing test failures, after adding the `-canonicalize` flag.

c1f2874 also removes the check for empty bodies. I don't think this is necessary. Let me know if you think I'm wrong.

### CODE REVIEW

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 1]
```diff
@@ -1345,6 +1387,70 @@ static LogicalResult verifyIfOp(IfOp ifOp) {
...
+///                                         calyx.enable @B
+static LogicalResult eliminateCommonTailEnable(IfOp ifOp,
+                                               PatternRewriter &rewriter) {
+  // Check if the branches exist.
+  if (!ifOp.thenRegionExists() || !ifOp.elseRegionExists())
+    return failure();
+
+  auto &thenOpStructureOp = ifOp.getThenBody()->front();
+  auto &elseOpStructureOp = ifOp.getElseBody()->front();
+  // TODO(circt/#1861): ParOps have less restrictive conditions.
```
Comment (cgyurgyik) [inline lib/Dialect/Calyx/CalyxOps.cpp:1419]: The `IfOp` region is going to be re-written quite a bit to account for #1861, so I wouldn't stare too hard.

File: include/circt/Dialect/Calyx/CalyxControl.td [hunk 2]
```diff
@@ -68,6 +68,7 @@ def IfOp : CalyxContainer<"if", [
 
   let assemblyFormat = "$cond (`with` $groupName^)? $thenRegion (`else` $elseRegion^)? attr-dict";
   let verifier = "return ::verify$cppClass(*this);";
+  let hasCanonicalizeMethod = true;
```
Comment (mortbopet) [inline include/circt/Dialect/Calyx/CalyxControl.td:71]: Why choose `hasCanonicalizeMethod` + explicit canonicalization function over the `hasCanonicalizer` + canonicalization patterns approach?

Comment (cgyurgyik) [reply include/circt/Dialect/Calyx/CalyxControl.td:71]: No strong reason. Any reason why you prefer one over the other?

Comment (mortbopet) [reply include/circt/Dialect/Calyx/CalyxControl.td:71]: I'd imagine `hasCanonicalizeMethod` is only to be preferred where:
1. An explicit order is required in the canonicalization method
2. There is a need to do complex transformations/logic which cannot be captured by the pattern rewrite driver.

Comment (cgyurgyik) [reply include/circt/Dialect/Calyx/CalyxControl.td:71]: Is this stated somewhere?

I mostly just read through [this](https://mlir.llvm.org/docs/Canonicalization/#canonicalizing-with-rewritepatterns), and it seemed the `hasCanonicalizeMethod` was the simplest of the two.

Comment (mikeurbach) [reply include/circt/Dialect/Calyx/CalyxControl.td:71]: I think `hasCanonicalizeMethod` makes sense to use whenever the situation is simple enough to fit into one simple pattern. If more flexibility is needed, `hasCanonicalizer` can be more general.

**Review by mikeurbach — APPROVED** [review]
nice, looks good to me.

**Review by mortbopet — APPROVED** [review]

**Pull request merged**